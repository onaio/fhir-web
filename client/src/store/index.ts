// going to use this store initializer instead of the one in
// @opensrp/store package, for some reason its not working
// wehn passed to the redux provider
import { connectReducer, getConnectedStore } from '@onaio/connected-reducer-registry';
import { gateKeeperReducer, gateKeeperReducerName } from '@onaio/gatekeeper';
import reducerRegistry, { combine, Registry } from '@onaio/redux-reducer-registry';
import session, { reducerName as sessionReducer } from '@onaio/session-reducer';
import {
  reducerName as keycloakUsersReducerName,
  reducer as keycloakUsersReducer,
} from '@opensrp/store';
import { Reducer } from 'redux';
import { RouterState } from 'connected-react-router';

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
  router: connectReducer as Reducer<RouterState>,
};

/** Add users reducer to registry */
defaultReducers[sessionReducer] = session;
defaultReducers[gateKeeperReducerName] = gateKeeperReducer;
// added this segment because registering from the user management
// package component is failing
defaultReducers[keycloakUsersReducerName] = keycloakUsersReducer;

/** The initial store for the reveal web app */
const store = getConnectedStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener((reducers) => {
  store.replaceReducer(combine(reducers));
});

export default store;
