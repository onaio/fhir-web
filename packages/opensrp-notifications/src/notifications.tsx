import { notification } from 'antd';

/**
 * Antd notification success popup
 *
 * @param {string} message - title for popup notification
 * @param {string} description - Description for popup notification
 */
export function sendSuccessNotification(message: string, description?: string) {
  notification.success({
    message: message,
    description: description,
  });
}

/**
 * Antd notification info popup
 *
 * @param {string} message - title for popup notification
 * @param {string} description - Description for popup notification
 */
export const sendInfoNotification = (message: string, description?: string): void => {
  notification.info({
    message: message,
    description: description,
  });
};

/**
 * Antd notification warning popup
 *
 * @param {string} message - title for popup notification
 * @param {string} description - Description for popup notification
 */
export const sendWarningNotification = (message: string, description?: string): void => {
  notification.warning({
    message: message,
    description: description,
  });
};

/**
 * Antd notification error popup
 *
 * @param {string} message - title for popup notification
 * @param {string} description - Description for popup notification
 */
export const sendErrorNotification = (message: string, description?: string): void => {
  notification.error({
    message: message,
    description: description,
  });
};
