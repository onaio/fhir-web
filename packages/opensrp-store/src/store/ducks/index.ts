/** users ducks modules: actions, actionCreators, reducer and selectors */
import { Dictionary } from '@onaio/utils';
import { keyBy, values } from 'lodash';
import intersect from 'fast_array_intersect';
import { AnyAction, Store } from 'redux';
import { createSelector, OutputParametricSelector } from 'reselect';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'keycloakUsers';

/** Interface for user json object */
export interface KeycloakUser {
  access?: {
    manageGroupMembership: boolean;
    view: boolean;
    mapRoles: boolean;
    impersonate: boolean;
    manage: boolean;
  };
  createdTimestamp?: number;
  disableableCredentialTypes?: string[];
  email?: string;
  emailVerified?: boolean;
  enabled?: boolean;
  firstName: string;
  id: string;
  lastName: string;
  notBefore?: number;
  requiredActions?: string[];
  totp?: boolean;
  username: string;
}

// actions

/** action type for fetching keycloak users */
export const KEYCLOAK_USERS_FETCHED = 'keycloak/reducer/users/userS_FETCHED';
/** action type for removing keycloak users */
export const REMOVE_KEYCLOAK_USERS = 'keycloak/reducer/users/REMOVE_KEYCLOAK_USERS';

/** interface action to add users to store */
export interface FetchKeycloakUsersAction extends AnyAction {
  usersById: Dictionary<KeycloakUser>;
  type: typeof KEYCLOAK_USERS_FETCHED;
}

/** Interface for removeusersAction */
export interface RemoveKeycloakUsersAction extends AnyAction {
  usersById: Dictionary<KeycloakUser>;
  type: typeof REMOVE_KEYCLOAK_USERS;
}

/** Create type for users reducer actions */
export type KeycloakUsersActionTypes =
  | FetchKeycloakUsersAction
  | RemoveKeycloakUsersAction
  | AnyAction;

// action Creators

/**
 * Fetch users action creator
 *
 * @param {Array} usersList - keycloak users
 * @returns {object} - the dispatcher to remove the user
 */
export const fetchKeycloakUsers = (usersList: KeycloakUser[] = []): FetchKeycloakUsersAction => {
  return {
    usersById: keyBy(usersList, (user: KeycloakUser) => user.id),
    type: KEYCLOAK_USERS_FETCHED,
  };
};

/** Remove users action creator
 *
 * @returns {object} - the dispatcher to remove the user
 */
export const removeKeycloakUsers = (): RemoveKeycloakUsersAction => {
  return {
    usersById: {},
    type: REMOVE_KEYCLOAK_USERS,
  };
};

// The reducer

/** interface for keycloak users state in redux store */
export interface KeycloakUsersState {
  usersById: Dictionary<KeycloakUser> | Dictionary;
}

/** Create an immutable keycloak users state */
export type ImmutableKeycloakUsersState = KeycloakUsersState &
  SeamlessImmutable.ImmutableObject<KeycloakUsersState>;

/** initial keycloak users state */
export const initialState: ImmutableKeycloakUsersState = SeamlessImmutable({
  usersById: {},
});

/**
 * the users reducer function
 *
 * @param {object} state - keycloak users states
 * @param {object} action - keycloak users actions
 * @returns {object} - the updated states
 */
export function reducer(
  state: ImmutableKeycloakUsersState = initialState,
  action: KeycloakUsersActionTypes
): ImmutableKeycloakUsersState {
  switch (action.type) {
    case KEYCLOAK_USERS_FETCHED:
      return SeamlessImmutable({
        ...state,
        usersById: { ...state.usersById, ...action.usersById },
      });
    case REMOVE_KEYCLOAK_USERS:
      return SeamlessImmutable({
        ...state,
        usersById: action.usersById,
      });

    default:
      return state;
  }
}

// Selectors

export interface KeycloakUsersFilters {
  id: string[] /** get all users whose ids appear in this array */;
  username?: string /** get users whose username includes text in the username field */;
}

// NON MEMOIZED SELECTORS

/**
 * Gets id from KeycloakUsersFilters
 *
 * @param {any} _ - the redux store
 * @param {Array} props - the keycloak users filters object
 * @returns {string} - the keycloak user id
 */
export const getUserIds = (_: Partial<Store>, props: KeycloakUsersFilters): string[] => props.id;

/**
 * Gets name from KeycloakUsersFilters
 *
 * @param {any} _ - the redux store
 * @param {object} props - the users filters object
 * @returns {string} - the keycloak user name
 */
export const getUsername = (_: Partial<Store>, props: KeycloakUsersFilters): string | undefined =>
  props.username;

/** returns all users in the store as values whose keys are their respective ids
 *
 * @param {any} state - the redux store
 * @returns {object} - users object as values, respective ids as keys
 */
export function getKeycloakUsersById(state: Partial<Store>): { [key: string]: KeycloakUser } {
  return (state as Dictionary)[reducerName].usersById;
}

/** gets keycloak users as an array of user objects
 *
 * @param {any} state - the redux store
 * @returns {Array} - an array of users objs
 */
export function getKeycloakUsersArray(state: Partial<Store>): KeycloakUser[] {
  return values(getKeycloakUsersById(state));
}

/**
 * Gets all practitioners whose name includes phrase given in name filter prop
 *
 * @returns {Array} - practitioners whose name includes phrase given in name filter prop
 */
export const getUsersByUsername = (): OutputParametricSelector<
  Partial<Store>,
  KeycloakUsersFilters,
  KeycloakUser[],
  (res1: KeycloakUser[], res2: string | undefined) => KeycloakUser[]
> =>
  createSelector(getKeycloakUsersArray, getUsername, (usersArray, username) =>
    username
      ? usersArray.filter((user: KeycloakUser) =>
          user.username.toLowerCase().includes(username.toLowerCase())
        )
      : usersArray
  );

// MEMOIZED SELECTORS

/**
 * Gets all users whose identifiers appear in ids filter prop value
 *
 * @returns {Array} - users whose identifiers appear in ids filter prop value
 */
export const getKeycloakUsersByIds = (): OutputParametricSelector<
  Partial<Store>,
  KeycloakUsersFilters,
  KeycloakUser[],
  (
    res1: {
      [key: string]: KeycloakUser;
    },
    res2: string[],
    res3: KeycloakUser[]
  ) => KeycloakUser[]
> =>
  createSelector(
    getKeycloakUsersById,
    getUserIds,
    getKeycloakUsersArray,
    (usersById, ids, usersArray) => {
      if (ids === undefined) {
        return usersArray;
      }
      if (ids.length > 0) {
        return ids.map((id: string) => usersById[id]);
      }
      return [];
    }
  );

/** practitioner array selector factory
 * aggregates response from all applied filters and returns results
 *
 * @returns {Array} - aggregates response from all applied filters and returns results
 */
export const makeKeycloakUsersSelector = (): OutputParametricSelector<
  Partial<Store>,
  KeycloakUsersFilters,
  KeycloakUser[],
  (res1: KeycloakUser[], res2: KeycloakUser[]) => KeycloakUser[]
> =>
  createSelector(getKeycloakUsersByIds(), getUsersByUsername(), (arr1, arr2) => {
    return intersect([arr1, arr2], JSON.stringify);
  });
