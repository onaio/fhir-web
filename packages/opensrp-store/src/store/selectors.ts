import { reducerName as sessionReducerName } from '@onaio/session-reducer';
import { Store } from 'redux';
import { Dictionary } from '@onaio/utils';
import { createSelector } from 'reselect';

/** get API Token from the Redux store
 *
 * @param {any} state - the redux store
 */
interface StoreState {
  session: {
    [key: string]: Dictionary;
  };
}

interface APIFilters {
  accessToken?: boolean;
  apiToken?: boolean;
  providerState?: boolean;
}

/** Fetch access token from store if access_token filter val is true
 *
 * @param {object} _ - redux store
 * @param {object} props - api filters object
 * @returns {boolean} returns boolean value
 */
export const fetchAccessToken = (_: Partial<Store>, props: APIFilters): boolean | null =>
  props.accessToken || null;

/** Fetch api token from store if api_token filter val is true
 *
 * @param {object} _ redux store
 * @param {object} props api filters object
 * @returns {boolean} returns boolean value
 */
export const fetchApiToken = (_: Partial<Store>, props: APIFilters): boolean | null =>
  props.apiToken || null;

/** Fetch oauth provider state from store if providerState filter val is true
 *
 * @param {Object} _ redux store
 * @param {Object} props api filters object
 * @returns {boolean} returns boolean value
 */
export const fetchOauthProviderState = (_: Partial<Store>, props: APIFilters): boolean | null =>
  props.providerState || null;

/** Gets extra data object from store
 *
 * @param {object} state - redux store
 * @returns {Object}
 */

export const getExtraData = (state: Partial<Store>): Dictionary =>
  (state as StoreState)[sessionReducerName].extraData;

/** API state selector
 *
 * @returns {Object} - the states
 */
export const makeAPIStateSelector = () =>
  createSelector(
    [fetchAccessToken, fetchApiToken, fetchOauthProviderState, getExtraData],
    (accessToken, apiToken, oauthState, extraData) => {
      if (accessToken) {
        return (extraData.oAuth2Data && extraData.oAuth2Data.access_token) || null;
      } else if (apiToken) {
        return extraData.api_token || null;
      } else if (oauthState) {
        return (extraData.oAuth2Data && extraData.oAuth2Data.state) || null;
      }
    }
  );
