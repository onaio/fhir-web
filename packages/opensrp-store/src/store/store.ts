import { connectReducer, getConnectedStore } from '@onaio/connected-reducer-registry';
import { gateKeeperReducer, gateKeeperReducerName } from '@onaio/gatekeeper';
import reducerRegistry, { combine, Registry } from '@onaio/redux-reducer-registry';
import session, { reducerName as sessionReducer } from '@onaio/session-reducer';
import { Reducer } from 'redux';
import { RouterState } from 'connected-react-router';
import { Dictionary } from '@onaio/utils';

/** declare globals interface */
declare global {
  interface Window {
    __PRELOADED_STATE__: Dictionary;
  }
}

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
  router: connectReducer as Reducer<RouterState>,
};

// Grab the state from a global variable injected into the server-generated HTML
const preloadedState = window.__PRELOADED_STATE__ || {};
// Allow the passed state to be garbage-collected
delete window.__PRELOADED_STATE__;

/** Add users reducer to registry */
defaultReducers[sessionReducer] = session;
defaultReducers[gateKeeperReducerName] = gateKeeperReducer;

/** The initial store for the reveal web app */
export const store = getConnectedStore(defaultReducers, preloadedState);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener((reducers) => {
  store.replaceReducer(combine(reducers));
});
