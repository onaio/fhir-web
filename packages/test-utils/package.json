{
  "name": "@opensrp/test-utils",
  "version": "0.0.0",
  "description": "",
  "main": "dist/index.js",
  "types": "dist/types",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": "https://github.com/opensrp/web",
  "scripts": {
    "test": "run -T test $INIT_CWD --verbose",
    "tsc": "run -T tsc",
    "lint": "run -T eslint ./**/*.{js,jsx,ts,tsx}",
    "copy": "run -T copyfiles -u 1 \"./src/**/*.{css,html}\" \"./dist/\"",
    "build": "run tsc && run transpile && run copy",
    "transpile": "run -T babel src -d dist --root-mode upward --extensions .ts,.tsx --ignore '**/*.test.ts,**/*.test.tsx,**/tests,**/__tests__'"
  },
  "jest": {
    "automock": false,
    "setupFiles": [
      "../../setupTests"
    ]
  },
  "bugs": {
    "url": "https://github.com/opensrp/web/issues"
  },
  "dependencies": {
    "@onaio/connected-reducer-registry": "^0.0.3",
    "@opensrp/pkg-config": "^0.0.9",
    "@opensrp/rbac": "workspace:^"
  },
  "peerDependencies": {
    "@opensrp/store": "^0.0.10",
    "@testing-library/react": "^14.1.2",
    "@testing-library/user-event": "^13.5.0",
    "react": "^17.0.0 ",
    "react-redux": "^7.2.1",
    "react-router": "^5.2.1",
    "react-router-dom": "^5.2.1"
  },
  "author": "OpenSRP Engineering",
  "license": "Apache-2.0"
}
