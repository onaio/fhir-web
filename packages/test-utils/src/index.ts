export * from './componentWrappers';
export * from './utils'