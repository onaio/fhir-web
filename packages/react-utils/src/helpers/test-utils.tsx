import { store } from '@opensrp/store';
import { Provider } from 'react-redux';
import React from 'react';
import { RoleContext, UserRole, FhirResources, Permit, IamResources } from '@opensrp/rbac';
import { history } from '@onaio/connected-reducer-registry';
import { Router } from 'react-router';

export const superUserRole = new UserRole([...FhirResources, ...IamResources], Permit.MANAGE);

export interface ContextProviderProps {
  children: JSX.Element;
}

export const ContextProvider = (props: ContextProviderProps) => {
  const { children } = props;
  return (
    <Provider store={store}>
      <Router history={history}>
        <RoleContext.Provider value={superUserRole}>{children}</RoleContext.Provider>
      </Router>
    </Provider>
  );
};
