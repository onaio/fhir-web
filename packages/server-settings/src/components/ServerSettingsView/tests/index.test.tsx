import { createBrowserHistory } from 'history';
import React from 'react';
import { shallow, mount } from 'enzyme';
import * as notifications from '@opensrp/notifications';
import { ServerSettingsView } from '..';
import {
  serverSettings,
  serverSettingsLevel1,
  securityAuthenticateEndpoint,
  locationHierarchyWithoutParent,
  locationHierarchyWithParent,
  serverSettingsSimilarToParent,
} from './fixtures';
import { act } from 'react-dom/test-utils';
import flushPromises from 'flush-promises';
import toJson from 'enzyme-to-json';
import fetch from 'jest-fetch-mock';
import { store } from '@opensrp/store';
import { QueryClient, QueryClientProvider } from 'react-query';
import { authenticateUser } from '@onaio/session-reducer';
import { Provider } from 'react-redux';
import { Router } from 'react-router';

const history = createBrowserHistory();

jest.mock('@opensrp/notifications', () => ({
  __esModule: true,
  ...Object.assign({}, jest.requireActual('@opensrp/notifications')),
}));

describe('activate mission', () => {
  beforeAll(() => {
    store.dispatch(
      authenticateUser(
        true,
        {
          email: 'bob@example.com',
          name: 'Bobbie',
          username: 'RobertBaratheon',
        },
        // eslint-disable-next-line @typescript-eslint/naming-convention
        { api_token: 'hunter2', oAuth2Data: { access_token: 'sometoken', state: 'abcde' } }
      )
    );
  });

  beforeEach(() => {
    fetch.resetMocks();
  });

  it('renders without crashing', async () => {
    const queryClient = new QueryClient();

    fetch.mockResponseOnce(JSON.stringify(securityAuthenticateEndpoint));
    fetch.mockResponseOnce(JSON.stringify(serverSettings));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <QueryClientProvider client={queryClient}>
            <ServerSettingsView
              baseURL="https://opensrp-stage.smartregister.org/opensrp/"
              restBaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/"
              v2BaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/v2/"
            />
          </QueryClientProvider>
        </Router>
      </Provider>
    );

    await act(async () => {
      await flushPromises();
    });
    wrapper.update();

    expect(fetch.mock.calls.map((res) => res[0])).toEqual([
      'https://opensrp-stage.smartregister.org/opensrp/security/authenticate',
      'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings?identifier=population_characteristics&locationId=02ebbc84-5e29-4cd5-9b79-c594058923e9&resolve=true&serverVersion=0',
    ]);

    expect(wrapper.find('Tree')).toBeTruthy();
    expect(wrapper.find('Tree').at(0).text()).toMatchInlineSnapshot(`"Uganda"`);
    expect(wrapper.find('Table')).toBeTruthy();
    expect(wrapper.find('Table').at(0).text()).toMatchInlineSnapshot(
      `"NameDescriptionSettingsInherited fromActionsAnaemia prevalence 20% or lowerThe proportion of pregnant women in the population with anaemia (haemoglobin level less than 11 g/dl) is 20% or lower.Yes-Anaemia prevalence 40% or higherThe proportion of pregnant women in the population with anaemia (haemoglobin level less than 11 g/dl) is 40% or higher.No-HIV incidence greater than 3 per 100 person-years in the absence of PrEPWomen in the population have a substantial risk of HIV infection. Substantial risk of HIV infection is provisionally defined as HIV incidence greater than 3 per 100 personâ€“years in the absence of pre-exposure prophylaxis (PrEP).No-HIV prevalence 5% or higherThe HIV prevalence in pregnant women in the population is 5% or higher.No-Hep B prevalence is intermediate (2% or higher) or high (5% or higher)The proportion of Hepatitis B surface antigen (HBsAg) seroprevalance in the general population is 2% or higher.No-Hep C prevalence is intermediate (2% or higher) or high (5% or higher)The proportion of Hepatitis C virus (HCV) antibody seroprevalence in the general population is 2% or higher. No-Low dietary calcium intakeWomen in the population are likely to have low dietary calcium intake (less than 900 mg of calcium per day).Yes-Malaria-endemic settingThis is a malaria-endemic setting.Yes-National Hep B ANC routine screening program establishedThere is a national Hepatitis B ANC routine screening program in place.Yes-Soil-transmitted helminth infection prevalence 20% or higherThe percentage of individuals in the general population infected with at least one species of soil-transmitted helminth is 20% or higher.No-12"`
    );
    wrapper.unmount();
  });

  it('Updates settings correctly when value is yes', async () => {
    const queryClient = new QueryClient({ defaultOptions: { queries: { retry: false } } });

    fetch.mockResponseOnce(JSON.stringify(securityAuthenticateEndpoint));
    fetch.mockResponseOnce(JSON.stringify(serverSettings));
    fetch.mockResponseOnce(JSON.stringify(locationHierarchyWithoutParent));
    fetch.mockResponseOnce(JSON.stringify({}));
    fetch.mockResponseOnce(JSON.stringify(serverSettings));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <QueryClientProvider client={queryClient}>
            <ServerSettingsView
              baseURL="https://opensrp-stage.smartregister.org/opensrp/"
              restBaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/"
              v2BaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/v2/"
            />
          </QueryClientProvider>
        </Router>
      </Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });
    const dropdown = wrapper.find('Dropdown').at(0);
    dropdown.simulate('click');
    const subMenu = shallow(<div>{dropdown.prop('overlay')}</div>);
    const yesBtn = subMenu.find('MenuItem').at(0);

    // click yes button
    yesBtn.simulate('click');

    await act(async () => {
      await flushPromises();
    });
    wrapper.update();

    expect(fetch.mock.calls.map((call) => [call[0], call[1]?.method])).toEqual([
      ['https://opensrp-stage.smartregister.org/opensrp/security/authenticate', 'GET'],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings?identifier=population_characteristics&locationId=02ebbc84-5e29-4cd5-9b79-c594058923e9&resolve=true&serverVersion=0',
        'GET',
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/heirarchy/ancestors/02ebbc84-5e29-4cd5-9b79-c594058923e9',
        'GET',
      ],
      ['https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings/5', 'PUT'],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings?identifier=population_characteristics&locationId=02ebbc84-5e29-4cd5-9b79-c594058923e9&resolve=true&serverVersion=0',
        'GET',
      ],
    ]);
    wrapper.unmount();
  });

  it('Updates settings correctly when value is no - and similar to parent settings', async () => {
    const queryClient = new QueryClient({ defaultOptions: { queries: { retry: false } } });

    fetch.mockResponseOnce(JSON.stringify(securityAuthenticateEndpoint));
    fetch.mockResponseOnce(JSON.stringify(serverSettingsSimilarToParent));
    fetch.mockResponseOnce(JSON.stringify(locationHierarchyWithParent));
    fetch.mockResponseOnce(JSON.stringify(serverSettingsSimilarToParent));
    fetch.mockResponseOnce(JSON.stringify({}));
    fetch.mockResponseOnce(JSON.stringify(serverSettingsSimilarToParent));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <QueryClientProvider client={queryClient}>
            <ServerSettingsView
              baseURL="https://opensrp-stage.smartregister.org/opensrp/"
              restBaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/"
              v2BaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/v2/"
            />
          </QueryClientProvider>
        </Router>
      </Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });
    const dropdown = wrapper.find('Dropdown').at(0);
    dropdown.simulate('click');
    const subMenu = shallow(<div>{dropdown.prop('overlay')}</div>);
    const yesBtn = subMenu.find('MenuItem').at(1);

    // click yes button
    yesBtn.simulate('click');

    await act(async () => {
      await flushPromises();
    });
    wrapper.update();

    expect(fetch.mock.calls.map((call) => [call[0], call[1]?.method])).toEqual([
      ['https://opensrp-stage.smartregister.org/opensrp/security/authenticate', 'GET'],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings?identifier=population_characteristics&locationId=02ebbc84-5e29-4cd5-9b79-c594058923e9&resolve=true&serverVersion=0',
        'GET',
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/location/heirarchy/ancestors/02ebbc84-5e29-4cd5-9b79-c594058923e9',
        'GET',
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings?identifier=population_characteristics&locationId=b652b2f4-a95d-489b-9e28-4629746db96a&resolve=true&serverVersion=0',
        'GET',
      ],
      ['https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings/5', 'DELETE'],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings?identifier=population_characteristics&locationId=02ebbc84-5e29-4cd5-9b79-c594058923e9&resolve=true&serverVersion=0',
        'GET',
      ],
    ]);
    wrapper.unmount();
  });

  it('Updates settings correctly when value is inherited', async () => {
    const queryClient = new QueryClient({ defaultOptions: { queries: { retry: false } } });

    fetch.mockResponseOnce(JSON.stringify(securityAuthenticateEndpoint));
    fetch.mockResponseOnce(JSON.stringify(serverSettings));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <QueryClientProvider client={queryClient}>
            <ServerSettingsView
              baseURL="https://opensrp-stage.smartregister.org/opensrp/"
              restBaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/"
              v2BaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/v2/"
            />
          </QueryClientProvider>
        </Router>
      </Provider>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });
    const dropdown = wrapper.find('Dropdown').at(0);
    dropdown.simulate('click');
    const subMenu = shallow(<div>{dropdown.prop('overlay')}</div>);
    const yesBtn = subMenu.find('MenuItem').at(2);

    // click yes button
    yesBtn.simulate('click');

    await act(async () => {
      await flushPromises();
    });
    wrapper.update();

    expect(fetch.mock.calls).toEqual([
      [
        'https://opensrp-stage.smartregister.org/opensrp/security/authenticate',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer sometoken',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings?identifier=population_characteristics&locationId=02ebbc84-5e29-4cd5-9b79-c594058923e9&resolve=true&serverVersion=0',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer sometoken',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings/5',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer sometoken',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'DELETE',
        },
      ],
      [
        'https://opensrp-stage.smartregister.org/opensrp/rest/v2/settings?identifier=population_characteristics&locationId=02ebbc84-5e29-4cd5-9b79-c594058923e9&resolve=true&serverVersion=0',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer sometoken',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
    ]);
    wrapper.unmount();
  });

  it('Table Changes on Tree Click', async () => {
    const queryClient = new QueryClient({ defaultOptions: { queries: { retry: false } } });

    fetch.once(JSON.stringify(securityAuthenticateEndpoint));
    fetch.mockResponse(JSON.stringify(serverSettings));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <QueryClientProvider client={queryClient}>
            <ServerSettingsView
              baseURL="https://opensrp-stage.smartregister.org/opensrp/"
              restBaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/"
              v2BaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/v2/"
            />
          </QueryClientProvider>
        </Router>
      </Provider>
    );

    await act(async () => {
      await flushPromises();
    });
    wrapper.update();

    await act(async () => {
      let treeholder = wrapper.find('div[className*="ant-tree-list-holder-inner"]');
      const firstNodeSwitcher = treeholder.find('span[className*="ant-tree-switcher"]').first();
      firstNodeSwitcher.simulate('click');
      await flushPromises();
      treeholder = wrapper.find('div[className*="ant-tree-list-holder-inner"]');
    });
    wrapper.update();

    await act(async () => {
      let treeholder = wrapper.find('div[className*="ant-tree-list-holder-inner"]');
      const secondNodeContent = treeholder.find('span[className*="ant-tree-title"]').last();
      secondNodeContent.simulate('click');
      treeholder = wrapper.find('div[className*="ant-tree-list-holder-inner"]');
    });
    wrapper.update();

    fetch.mockResponse(JSON.stringify(serverSettingsLevel1));

    await act(async () => {
      await flushPromises();
    });
    wrapper.update();

    // test snapshot of expanded tree
    expect(wrapper.find('Tree').at(0).text()).toMatchInlineSnapshot(`"UgandaKampalaKCCA"`);
    wrapper.unmount();
  });

  it('handles errors', async () => {
    const queryClient = new QueryClient({ defaultOptions: { queries: { retry: false } } });
    fetch.mockRejectOnce(new Error('API is down'));
    const notificationErrorMock = jest.spyOn(notifications, 'sendErrorNotification');

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <QueryClientProvider client={queryClient}>
            <ServerSettingsView
              baseURL="https://opensrp-stage.smartregister.org/opensrp/"
              restBaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/"
              v2BaseURL="https://opensrp-stage.smartregister.org/opensrp/rest/v2/"
            />
          </QueryClientProvider>
        </Router>
      </Provider>
    );

    // Loader should be displayed
    expect(toJson(wrapper.find('.ant-spin'))).toBeTruthy();

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });
    expect(notificationErrorMock).toHaveBeenCalledWith('An error occurred');

    // broken page as well
    expect(wrapper.text()).toMatchInlineSnapshot(`"ErrorSomething went wrongGo backGo home"`);
  });
});
