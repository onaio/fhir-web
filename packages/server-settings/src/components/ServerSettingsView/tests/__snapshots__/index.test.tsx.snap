// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activate mission Updates settings correctly when value is inherited 1`] = `null`;
